### ISSUE TITLE
Misc build errors for amalgamated build on legacy iOS target for version 2.11

### ISSUE THREAD

**Issue by pgp** [randombit/botan#2045]
Hi everyone,
I've recently updated the amalgamated sources from version 2.9 to version 2.11 for one of my projects https://github.com/pgp/XFilesRootHelper , and there are three kinds of errors that did not appear with version 2.9 (and that I solved by manually patching the amalgamated source and headers with suitable modifications).

The build target is iOS 6, the amalgamation generation script can be found [here](https://github.com/pgp/XFilesRootHelper/blob/master/genBotanAmalgamations.sh#L48) , while the cmake build script is [here](https://github.com/pgp/XFilesRootHelper/blob/master/CMAKE_ios/CMakeLists.txt) and the build instructions are at the end of the README file [here](https://github.com/pgp/XFilesRootHelper/blob/master/README.md)

More precisely, the errors are:

- one occurrence of the thread_local modifier, which is not available for iOS < 8
(it depends on the macro BOTAN_TARGET_OS_HAS_THREADS whose existence is checked in [botan_all.h](https://github.com/pgp/XFilesRootHelper/blob/master/botanAm/ios/arm/botan_all.h) , which is presumably defined and hence the modifier is enabled); the only usage of the modifier is in [botan_all.cpp](https://github.com/pgp/XFilesRootHelper/blob/master/botanAm/ios/arm/botan_all.cpp) :
  ```cpp
  static BOTAN_THREAD_LOCAL CPUID::CPUID_Data g_cpuid;
  ```
  patched by maually removing the modifier along with the BOTAN_THREAD_LOCAL macro definition;

- Incomplete type struct stat, caused by the conditional include in [botan_all.cpp](https://github.com/pgp/XFilesRootHelper/blob/master/botanAm/ios/arm/botan_all.cpp) of the following:
	```cpp
	#elif defined(BOTAN_TARGET_OS_HAS_DEV_RANDOM)
	#include <sys/types.h>
	#include <sys/stat.h>
	#include <fcntl.h>
	#endif
	```
   and patched by forcing the above includes directly in botan_all.h

- Undeclared identifier MAP_ANONYMOUS in [botan_all.cpp](https://github.com/pgp/XFilesRootHelper/blob/master/botanAm/ios/arm/botan_all.cpp)
  patched by replacing it with the less portable MAP_ANON
  

Now, my question is: are there any configure flags for the ios+armv7 target in order to avoid these build errors without manual patching?
Should I simply define/undefine the responsible macros with the -D compiler switch?
(I assume you are maintaining compatibility only with recent iOS versions, so I'm not necessarily asking for a fix for retrocompatibility, if that conflicts with your current requirements/roadmap/whatever)

**Comment by randombit** [comment]
My understanding is that Apple really only supports the latest version of iOS for security updates, and as a general project policy we only aim to support OSes which are supported wrt security issues by their vendor. We do make some accommodations based on market share (eg we will continue to support Windows 7 for some time in the future even though it goes EOL in January, because it still has 30+% market share). However I think that doesn't apply so much for iOS < 10 since most stats I can find online indicate these versions are quite rarely used.

That said, the `MAP_ANON` issue is actually a bug in our side; we previously had handling for that and it went missing when a relevant part of the code was rewritten. Its just a couple of lines of code so that's an easy fix on our side. But there is no way for you to fix it on your side short of -DMAP_ANONYMOUS=MAP_ANON or the like.

Similarly, the `stat` thing sounds like a bug on our part. Can you trace down which part of the code was using `stat` that saw an error until you hacked the includes? Just fn name should be enough for me to trace it down.

Re the thread local thing, you *can* disable it at build time with --without-os-features=threads but that is a big hammer which disables all threading support (use of threads for perf improvements, all internal locking, etc). But I'll look at splitting this out so in a future release you can use --without-os-feature=thread_local and just disable the thread local usage without also disabling locking and the like. In the meantime just patching it out is probably best.

**Comment by pgp** [comment]
Thank you for your efforts!
As for the <sys/stat.h> include, it seems it doesn't belong to any particular function or namespace, you can find it in the code block of lines 65826 to 65845 of [botan_all.cpp](https://github.com/pgp/XFilesRootHelper/blob/master/botanAm/ios/arm/botan_all.cpp) - actually, it is the only occurrence of including sys/stat.h in the whole source file.

**Comment by randombit** [comment]
Ok now I see the problem with the stat includes. I'll fix this tonight.

**Comment by randombit** [comment]
All three issues should be addressed now in master. You would still need to use `--without-os-feature=thread_local` to disable `thread_local` usage. Thanks for reporting these issues. These fixes will be included in 2.12, eta early October.
